From e161b5b2ec61a72d616b77fb567f2b37dde12eef Mon Sep 17 00:00:00 2001
From: CIQ Kernel Automation <ciq_kernel_automation@ciq.com>
Date: Thu, 30 Apr 2026 07:08:12 +0000
Subject: [PATCH] mm/damon/sysfs: cleanup attrs subdirs on context dir setup
 failure

jira VULN-176247
cve CVE-2026-23144
commit-author SeongJae Park <sj@kernel.org>
commit 9814cc832b88bd040fc2a1817c2b5469d0f7e862

When a context DAMON sysfs directory setup is failed after setup of attrs/
directory, subdirectories of attrs/ directory are not cleaned up.  As a
result, DAMON sysfs interface is nearly broken until the system reboots,
and the memory for the unremoved directory is leaked.

Cleanup the directories under such failures.

Link: https://lkml.kernel.org/r/20251225023043.18579-3-sj@kernel.org
Fixes: c951cd3b8901 ("mm/damon: implement a minimal stub for sysfs-based DAMON interface")
	Signed-off-by: SeongJae Park <sj@kernel.org>
	Cc: chongjiapeng <jiapeng.chong@linux.alibaba.com>
	Cc: <stable@vger.kernel.org> # 5.18.x
	Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
(cherry picked from commit 9814cc832b88bd040fc2a1817c2b5469d0f7e862)
	Signed-off-by: CIQ Kernel Automation <ciq_kernel_automation@ciq.com>
---
 mm/damon/sysfs.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/mm/damon/sysfs.c b/mm/damon/sysfs.c
index 07e5f1bdf025..f839406cc091 100644
--- a/mm/damon/sysfs.c
+++ b/mm/damon/sysfs.c
@@ -1772,7 +1772,7 @@ static int damon_sysfs_context_add_dirs(struct damon_sysfs_context *context)
 
 	err = damon_sysfs_context_set_targets(context);
 	if (err)
-		goto put_attrs_out;
+		goto rmdir_put_attrs_out;
 
 	err = damon_sysfs_context_set_schemes(context);
 	if (err)
@@ -1782,7 +1782,8 @@ static int damon_sysfs_context_add_dirs(struct damon_sysfs_context *context)
 put_targets_attrs_out:
 	kobject_put(&context->targets->kobj);
 	context->targets = NULL;
-put_attrs_out:
+rmdir_put_attrs_out:
+	damon_sysfs_attrs_rm_dirs(context->attrs);
 	kobject_put(&context->attrs->kobj);
 	context->attrs = NULL;
 	return err;
-- 
2.47.3

